<a href="https://colab.research.google.com/github/peremartra/Rearchitecting-LLMs/blob/main/CH02/CH02_NB02_Knowledge_Recovery.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Rearchitecting LLMs**
## **Structural techniques for efficient models**

### **Chapter 2: Knowledge Recovery via Distillation**
by [Pere Martra](https://github.com/peremartra)

[![LinkedIn](https://img.shields.io/badge/LinkedIn-0077B5?style=flat&logo=linkedin&logoColor=white)](https://www.linkedin.com/in/pere-martra/) [![GitHub](https://img.shields.io/badge/GitHub-100000?style=flat&logo=github&logoColor=white)](https://github.com/peremartra) [![Hugging Face](https://img.shields.io/badge/🤗%20Hugging%20Face-blue)](https://huggingface.co/oopere)

___

**Colab Environment:** GPU T4  
- **Models recommended:** google/gemma-3-270m  
- **Tested with:** meta-llama/Llama-3.2-1B using a **GPU A100**

---

Welcome to the second part of your first model rearchitecture. In the previous notebook, we successfully performed depth pruning by removing layers from our model, achieving significant efficiency gains but experiencing expected performance degradation.

Now we'll complete the optimization cycle by recovering the lost knowledge through Knowledge Distillation (KD). This notebook demonstrates how a pruned model can learn to mimic the behavior of its original, unpruned version.

**What we'll accomplish:**
- **Load our models**: Original (teacher) and pruned (student) from the previous notebook
- **Prepare recovery data**: Prepare a general-purpose dataset (SlimPajama) to facilitate the knowledge transfer.
- **Apply Knowledge Distillation**: Train the pruned model to recover lost capabilities
- **Measure recovery**: Quantify how much performance we can restore

This notebook bridges the gap between "breaking" a model (pruning) and "fixing" it (recovery), completing your first end-to-end model tailoring workflow.

**Previous notebook:** [CH02_NB01_Depth_pruning_evaluation](https://github.com/peremartra/Rearchitecting-LLMs/blob/main/CH02/CH02_NB01_Depth_pruning_evaluation.ipynb)
**Connection:** We'll use the exact same models and evaluation framework to measure our recovery success.


In [16]:
# Libraries
# Install required packages
!pip install -q transformers torch optipfair datasets accelerate sentencepiece lm-eval

In [17]:
import torch
import torch.nn as nn
from transformers import AutoModelForCausalLM, AutoTokenizer
from torch.optim import AdamW
from optipfair import prune_model
from datasets import load_dataset
from torch.nn import functional as F
from torch.utils.data import DataLoader
from lm_eval import evaluator
from lm_eval.models.huggingface import HFLM
import time
import json
import numpy as np
from typing import Dict, List, Any
import copy, random, os
from transformers import set_seed

# Check device
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f"Using device: {device}")
if torch.cuda.is_available():
    print(f"GPU: {torch.cuda.get_device_name(0)}")

Using device: cuda
GPU: NVIDIA L4


In [18]:
def set_reproducibility(seed=42):
    # 1. Seed for Python and basic libraries
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)

    # 2. Seed for PyTorch (CPU and GPU)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed) # For multi-GPU

    # 3. Configure cuDNN to be deterministic
    # Note: This may slow down training slightly
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

    # 4. Seed for Transformers (Hugging Face)
    set_seed(seed)

    print(f"✅ Seed {seed} stablished.")

set_reproducibility(42)

✅ Seed 42 stablished.


# 2.4 Recovering knowledge with distillation

## Load Libraries & Models.

In [ ]:
# Model configuration (consistent with previous notebook)
MODEL_NAME = "google/gemma-3-270m"
#MODEL_NAME = "meta-llama/Llama-3.2-1B"
MAX_NEW_TOKENS = 50
LAYERS_TO_REMOVE = 2 #Try removing 4, 6, or even 8 layers
TEST_PROMPT = "Paris is the capital of"

print(f"Loading base model: {MODEL_NAME}")

# Load the original model (this will be our TEACHER)
teacher_model = AutoModelForCausalLM.from_pretrained(
    MODEL_NAME,
    device_map="auto"
)

tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token

# Clean generation config (same as previous notebook)
from transformers import GenerationConfig
clean_config = GenerationConfig(
    max_length=teacher_model.generation_config.max_length,
    pad_token_id=teacher_model.generation_config.pad_token_id,
    eos_token_id=teacher_model.generation_config.eos_token_id,
    do_sample=False,
    num_beams=1,
    early_stopping=False
)
teacher_model.generation_config = clean_config

In [20]:
original_params = sum(p.numel() for p in teacher_model.parameters())
print(f"Teacher model parameters: {original_params:,}")
print(f"Teacher model layers: {len(teacher_model.model.layers)}")

Teacher model parameters: 268,098,176
Teacher model layers: 18


### Create Student Model with optiPfair


In [21]:
# Create the STUDENT model by pruning (same process as previous notebook)
print(f"\nCreating student model by removing {LAYERS_TO_REMOVE} layers...")

# Apply depth pruning using optipfair
student_model = prune_model(
    model=copy.deepcopy(teacher_model),
    pruning_type="DEPTH",
    num_layers_to_remove=LAYERS_TO_REMOVE,
    layer_selection_method="last",
    show_progress=True
)


Creating student model by removing 2 layers...


Removing layers: 100%|██████████| 18/18 [00:00<00:00, 194080.90it/s]


In [22]:
student_params = sum(p.numel() for p in student_model.parameters())
param_reduction = (original_params - student_params) / original_params

print(f"Student model parameters: {student_params:,}")
print(f"Parameter reduction: {param_reduction:.1%}")
print(f"Student model layers: {len(student_model.model.layers)}")

student_model.gradient_checkpointing_enable()

Student model parameters: 256,950,912
Parameter reduction: 4.2%
Student model layers: 16


## Support Functions & Basic Test

In [23]:
def model_evaluation(model_obj, tokenizer, tasks, limit=100):
    """
    Runs lm-eval on a PyTorch model object already in memory.

    Args:
        model_obj: The PyTorch model object to evaluate.
        tokenizer: The tokenizer object.
        tasks (list): A list of task names.
        limit (int): The number of samples per task.
    """
    print(f"Starting lm-eval on model '{model_obj.config._name_or_path}' for tasks: {tasks}")

    # Wrap the local model object and tokenizer for lm-eval
    model_wrapper = HFLM(
        pretrained=model_obj,
        tokenizer=tokenizer,
        device=str(device)
    )

    results = evaluator.simple_evaluate(
        model=model_wrapper,
        tasks=tasks,
        num_fewshot=0,
        limit=limit,
        device=str(device),
    )

    # Format results for clean display
    formatted_results = {}
    for task_name, res in results["results"].items():
        # Look for accuracy ('acc') first, then perplexity ('ppl')
        if 'acc,none' in res:
            metric_val = res.get('acc,none', 0)
        elif 'ppl,none' in res:
             metric_val = res.get('ppl,none', 0)
        else:
            metric_val = 0 # Fallback

        formatted_results[task_name] = f"{metric_val:.4f}"

    print(json.dumps(formatted_results, indent=2))
    return formatted_results

In [24]:
# Quick baseline test - confirm degradation from previous notebook
def generate_text(model, tokenizer, prompt: str, max_new_tokens: int = MAX_NEW_TOKENS) -> str:
    """Generate text with the model (same function as previous notebook)"""
    inputs = tokenizer(prompt, return_tensors='pt').to(device)
    with torch.no_grad():
        outputs = model.generate(
            inputs['input_ids'],
            attention_mask=inputs['attention_mask'],
            max_new_tokens=max_new_tokens,
            num_return_sequences=1,
            pad_token_id=tokenizer.pad_token_id,
            do_sample=False,
            num_beams=3,
            early_stopping=True,
            no_repeat_ngram_size=2
        )
    return tokenizer.decode(outputs[0], skip_special_tokens=True)

# Test both models with the same prompt
print(f"\n--- Baseline Test: '{TEST_PROMPT}' ---")
teacher_output = generate_text(teacher_model, tokenizer, TEST_PROMPT)
student_output = generate_text(student_model, tokenizer, TEST_PROMPT)

print(f"Teacher: '{teacher_output}'")
print(f"Student: '{student_output}'")
print("\nReady for knowledge recovery...")


--- Baseline Test: 'Paris is the capital of' ---
Teacher: 'Paris is the capital of France. It is located in the middle of the country and has a population of about 8 million people. The city is known for its beautiful architecture, rich history, and diverse culture. Here are some things to do in Paris: 1.'
Student: 'Paris is the capital of France and one of the largest cities in Europe. It occupies approximately 2.5 million hectares of land surrounded by mountains and forests. Parisians love to travel abroad because they enjoy sightseeing tours abroad. Tourists visiting Paris visit museums, monuments, theaters,'

Ready for knowledge recovery...


##Knowledge Distillation Process

Before we begin the recovery process, let's review the performance impact from our depth pruning (results from the previous notebook):

### Gemma-3-270m: Depth Pruning Impact

| Metric | Original Model | Pruned Model (-2 Layers) | Pruned Model (-3 Layers) | Change (-2L) | Change (-3L) |
|:-------|:---------------|:-------------------------|:-------------------------|:-------------|:-------------|
| **Parameters** | 268,098,176 | 256,950,912 | 251,377,280 | **-4.16%** | **-6.24%** |
| **Inference Time** | 10.324s | 4.322s | - | **+58.1%** | - |
| **arc_easy** (acc) | 0.5500 | 0.4600 | 0.3800 | -16.36% | -30.91% |
| **winogrande** (acc) | 0.6000 | 0.4800 | 0.5400 | -20.00% | -10.00% |
| **boolq** (acc) | 0.6600 | 0.4500 | 0.7000 | -31.82% | **+6.06%** |
| **lambada_openai** (acc) | 0.4200 | 0.3400 | 0.1800 | -19.05% | -57.14% |

**Trade-off**: 58.1% faster inference + 4.16% parameter reduction vs. accuracy degradation across benchmarks.

---

### LLaMA-3.2-1B: Depth Pruning Impact

| Metric | Original Model | Pruned Model (-2 Layers) | Change |
|:-------|:---------------|:-------------------------|:-------|
| **Parameters** | 1,235,814,400 | 1,117,171,392 | **-9.84%** |
| **Inference Time** | 6.635s | 4.752s | **+28.4%** |
| **arc_easy** (acc) | 0.6600 | 0.4800 | -27.27% |
| **winogrande** (acc) | 0.6000 | 0.5400 | -10.00% |
| **boolq** (acc) | 0.6700 | 0.7000 | **+4.48%** |
| **lambada_openai** (acc) | 0.5700 | 0.1700 | -70.18% |

**Interesting observation**: LLaMA-3.2-1B shows a curious **improvement** in boolq (+4.48%) after pruning, suggesting some layers may have been introducing noise for this specific task.

**Our KD goal**: Recover the lost performance while maintaining the efficiency gains from pruning.

## Dataset Preparation


In [ ]:
# Load SlimPajama dataset in streaming mode for efficiency
print("Loading SlimPajama dataset...")
"""
The original slimpajama-627B dataset seems to have disappeared from
Hugging Face, I've replaced it with the slimpajama-6B dataset temporarily,
in case SlimPajama-627B is permanently discontinued both the book and
the notebook text will be modified to adapt to the
results obtained with a new dataset."""

dataset = load_dataset(
    "DKYoon/SlimPajama-6B",  #"cerebras/SlimPajama-627B",
    split="train",
    streaming=True
)

# Take a representative subset for our recovery process
#    You can reduce the number os samples to reduce the execution time
#    but you'll see some reduction in lambada benchmark
#    due to small number of examples
RECOVERY_SAMPLES = 15000
print(f"Selecting {RECOVERY_SAMPLES:,} samples for knowledge recovery...")

# Use streaming dataset's take method - much more efficient!
distillation_dataset = dataset.take(RECOVERY_SAMPLES)

print(f"✓ Streaming dataset ready: {RECOVERY_SAMPLES:,} samples")

In [26]:
def tokenize_for_kd(examples, max_length=128):
    """
    Tokenize text for Knowledge Distillation.
    Shorter sequences work better for KD as both teacher and student
    need to fit in memory simultaneously.
    """
    if isinstance(examples, dict):
        texts = examples["text"]
    else:
        texts = [ex["text"] for ex in examples]

    tokenized = tokenizer(
        texts,
        padding="max_length",
        truncation=True,
        max_length=max_length,
        return_tensors="pt"
    )

    # For language modeling, labels = input_ids
    return {
        "input_ids": tokenized["input_ids"],
        "attention_mask": tokenized["attention_mask"],
        "labels": tokenized["input_ids"].clone()
    }

# Process the recovery dataset
print("Tokenizing recovery dataset...")

# Convert streaming dataset to list for tokenization
recovery_samples = list(distillation_dataset)
tokenized_data = tokenize_for_kd(recovery_samples)

print(f"✓ Tokenized {len(recovery_samples):,} samples")

# Convert to format suitable for DataLoader
class KDDataset(torch.utils.data.Dataset):
    def __init__(self, input_ids, attention_mask, labels):
        self.input_ids = input_ids
        self.attention_mask = attention_mask
        self.labels = labels

    def __len__(self):
        return len(self.input_ids)

    def __getitem__(self, idx):
        return {
            'input_ids': self.input_ids[idx],
            'attention_mask': self.attention_mask[idx],
            'labels': self.labels[idx]
        }
g = torch.Generator()
g.manual_seed(42)

kd_dataset = KDDataset(
    tokenized_data["input_ids"],
    tokenized_data["attention_mask"],
    tokenized_data["labels"]
)

# Create DataLoader for training
kd_dataloader = DataLoader(
    kd_dataset,
    batch_size=8,  # Small batch size due to memory constraints with two models
    shuffle=True,
    generator=g
)

print(f"✓ Knowledge Distillation DataLoader ready: {len(kd_dataloader)} batches")

Tokenizing recovery dataset...
✓ Tokenized 15,000 samples
✓ Knowledge Distillation DataLoader ready: 1875 batches


In [27]:
# Move models to device and set appropriate modes
teacher_model.to(device)
student_model.to(device)

# Teacher stays in eval mode - we don't train it
teacher_model.eval()

# Student will be trained
student_model.train()

# KD Hyperparameters
TEMPERATURE = 2.0      # Softens probability distributions
ALPHA = 1.0           # Weight for distillation loss
NUM_EPOCHS = 3        # Conservative for demo
LEARNING_RATE = 1e-5  # Lower LR for stability
ACCUMULATION_STEPS = 4  # Effective batch size = 4 * 8 = 32

# Optimizer for student model only
optimizer = AdamW(student_model.parameters(), lr=LEARNING_RATE)

print(f"Knowledge Distillation Configuration:")
print(f"  Temperature: {TEMPERATURE}")
print(f"  Alpha: {ALPHA}")
print(f"  Epochs: {NUM_EPOCHS}")
print(f"  Learning Rate: {LEARNING_RATE}")
print(f"  Effective Batch Size: {4 * ACCUMULATION_STEPS}")

Knowledge Distillation Configuration:
  Temperature: 2.0
  Alpha: 1.0
  Epochs: 3
  Learning Rate: 1e-05
  Effective Batch Size: 16


In [28]:
print(f"\n Starting Knowledge Distillation Training...")
print(f"Training student model to mimic teacher behavior \n")

for epoch in range(NUM_EPOCHS):
  student_model.train()
  total_loss = 0
  num_batches = 0
  for batch_idx, batch in enumerate(kd_dataloader):
    # ###Step 1: Move batch to device###
    input_ids = batch['input_ids'].to(device)
    attention_mask = batch['attention_mask'].to(device)

    # Move teacher model to device, perform inference, and move back to CPU
    teacher_model.to(device)

    # ###Step 2: The Master is asked to generate its logits###
    with torch.no_grad():
      teacher_outputs = teacher_model(
        input_ids=input_ids,
        attention_mask=attention_mask
      )
      teacher_logits = teacher_outputs.logits / TEMPERATURE

    # Moving teacher model to CPU to save memory.
    teacher_model.cpu()
    torch.cuda.empty_cache()

    # ###Step 3: Get Student's "thoughts" (logits) with training enabled###
    # Student inference (with gradients)
    student_outputs = student_model(
      input_ids=input_ids,
      attention_mask=attention_mask
    )
    student_logits = student_outputs.logits / TEMPERATURE

    # Compute Knowledge Distillation loss
    teacher_probs = F.softmax(teacher_logits, dim=-1)
    student_log_probs = F.log_softmax(student_logits, dim=-1)

    # ###Step 4:
    # KL Divergence loss
    kd_loss = F.kl_div(
      student_log_probs,
      teacher_probs,
      reduction='batchmean'
    )

    # Scale loss for gradient accumulation
    loss = kd_loss / ACCUMULATION_STEPS
    loss.backward()

    # Gradient accumulation
    if (batch_idx + 1) % ACCUMULATION_STEPS == 0 or (batch_idx + 1) == len(kd_dataloader):
      optimizer.step()
      optimizer.zero_grad()

    total_loss += loss.item() * ACCUMULATION_STEPS
    num_batches += 1
    # Progress update

    if (batch_idx + 1) % 100 == 0:
      avg_loss = total_loss / num_batches
      print(f'Epoch {epoch + 1}/{NUM_EPOCHS} | Batch {batch_idx + 1} | Loss: {avg_loss:.4f}')

  # Epoch summary
  avg_epoch_loss = total_loss / num_batches
  print(f"Epoch {epoch + 1}/{NUM_EPOCHS} | Average Loss: {avg_epoch_loss:.4f}")

print(f"\n Knowledge Distillation completed!")


🎓 Starting Knowledge Distillation Training...
Training student model to mimic teacher behavior 



`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`.


Epoch 1/3 | Batch 100 | Loss: 125.4800
Epoch 1/3 | Batch 200 | Loss: 103.9250
Epoch 1/3 | Batch 300 | Loss: 92.9325
Epoch 1/3 | Batch 400 | Loss: 86.0400
Epoch 1/3 | Batch 500 | Loss: 80.8645
Epoch 1/3 | Batch 600 | Loss: 77.2633
Epoch 1/3 | Batch 700 | Loss: 74.1511
Epoch 1/3 | Batch 800 | Loss: 71.5134
Epoch 1/3 | Batch 900 | Loss: 69.3964
Epoch 1/3 | Batch 1000 | Loss: 67.5700
Epoch 1/3 | Batch 1100 | Loss: 66.0175
Epoch 1/3 | Batch 1200 | Loss: 64.5921
Epoch 1/3 | Batch 1300 | Loss: 63.2504
Epoch 1/3 | Batch 1400 | Loss: 62.1545
Epoch 1/3 | Batch 1500 | Loss: 61.0697
Epoch 1/3 | Batch 1600 | Loss: 60.2373
Epoch 1/3 | Batch 1700 | Loss: 59.3365
Epoch 1/3 | Batch 1800 | Loss: 58.4996
Epoch 1/3 | Average Loss: 57.9501
Epoch 2/3 | Batch 100 | Loss: 42.8700
Epoch 2/3 | Batch 200 | Loss: 42.9875
Epoch 2/3 | Batch 300 | Loss: 43.2792
Epoch 2/3 | Batch 400 | Loss: 43.2600
Epoch 2/3 | Batch 500 | Loss: 43.3635
Epoch 2/3 | Batch 600 | Loss: 43.0800
Epoch 2/3 | Batch 700 | Loss: 42.9743
Epoch

## Basic Test generation.

In [29]:
# Set student model to evaluation mode
teacher_model.to(device)
student_model.eval()
# Test with the same prompt used in baseline
print(f"--- Qualitative Test: '{TEST_PROMPT}' ---")

# Generate with all three models for comparison
teacher_output = generate_text(teacher_model, tokenizer, TEST_PROMPT)
student_baseline_output = generate_text(student_model, tokenizer, TEST_PROMPT)

print(f"Teacher (Original):    '{teacher_output}'")
print(f"Student (Post-KD):     '{student_baseline_output}'")

--- Qualitative Test: 'Paris is the capital of' ---
Teacher (Original):    'Paris is the capital of France. It is located in the middle of the country and has a population of about 8 million people. The city is known for its beautiful architecture, rich history, and diverse culture. Here are some things to do in Paris: 1.'
Student (Post-KD):     'Paris is the capital of France and one of the most famous cities in Europe. It is also the largest city in France, with a population of more than 8 million people. Paris is located on the river Seine, which flows through the city. The city is surrounded by'


## Evaluation

In [ ]:
# Define the benchmark suite for our diagnostic
benchmark_tasks = ['arc_easy', 'winogrande', 'hellaswag', 'lambada_openai']
student_recovered_results = model_evaluation(student_model, tokenizer, benchmark_tasks, limit=100)

In [31]:
student_recovered_results

{'arc_easy': '0.4900',
 'hellaswag': '0.3900',
 'lambada_openai': '0.2200',
 'winogrande': '0.5300'}

10:47
{ "arc_easy": "0.4600", "hellaswag": "0.3900", "lambada_openai": "0.3300", "winogrande": "0.5600" }

Limit 1000 Seed 42
{
  "arc_easy": "0.5270",
  "hellaswag": "0.3630",
  "lambada_openai": "0.3680",
  "winogrande": "0.5460"
}


## Complete Knowledge Recovery Analysis

## Results with 2 layers pruned

| **Metric** | **Teacher (Original)** | **Student (-2L pruned)** | **Student (-2L, 500 samples)** | **Student (-2L, 15K samples)** | **Final Performance vs Original** |
|------------|------------------------|--------------------------|--------------------------------|--------------------------------|-----------------------------------|
| **Parameters** | 268,098,176 | 256,950,912 | 256,950,912 | 256,950,912 | **-4.16%** |
| **arc_easy** (acc) | 0.5500 | 0.4600 | 0.4700 | **0.5200** | **94.5%** of original |
| **hellaswag** (acc) | 0.4200 | - | - | **0.4100** | **97.6%** of original |
| **lambada_openai** (acc) | 0.4200 | 0.3400 | 0.2600 | **0.3900** | **92.9%** of original |
| **winogrande** (acc) | 0.6000 | 0.4800 | 0.5800 | **0.5800** | **96.7%** of original |

## Results with 3 layers pruned
| Metric | Teacher (Original) | Student (-3L pruned) | Student (-3L, 5K samples) | Final Performance vs Original |
|:-------|:-------------------|:---------------------|:--------------------------|:------------------------------|
| **Parameters** | 268,098,176 | 251,377,280 | 251,377,280 | **-6.24%** |
| **arc_easy** (acc) | 0.5500 | 0.3800 | **0.4700** | **85.5%** of original |
| **winogrande** (acc) | 0.6000 | 0.5400 | **0.5800** | **96.7%** of original |
| **boolq** (acc) | 0.6600 | 0.7000 | **0.6700** | **101.5%** of original |
| **lambada_openai** (acc) | 0.4200 | 0.1800 | **0.3300** | **78.6%** of original |

## Results with 4 layers pruned

| **Metric** | **Teacher (Original)** | **Student (-4L pruned)** | **Student (-4L, 5K samples)** | **Student (-4L, 15K samples)** | **Final Performance vs Original** |
|------------|------------------------|--------------------------|--------------------------------|--------------------------------|-----------------------------------|
| **Parameters** | 268,098,176 | 245,803,648 | 245,803,648 | 245,803,648 | **-8.31%** |
| **arc_easy** (acc) | 0.5500 | 0.4100 | **0.4800** | **0.4300** | **78.2%** of original |
| **hellaswag** (acc) | 0.4200 | 0.3100 | **0.3800** | **0.3600** | **85.7%** of original |
| **lambada_openai** (acc) | 0.4200 | 0.2100 | **0.2700** | **0.3300** | **78.6%** of original |
| **winogrande** (acc) | 0.6000 | 0.5300 | **0.5800** | **0.6000** | **100.0%** of original |

### LLaMA-3.2-1B (-2 Layers) Knowledge Recovery

| Metric | Teacher (Original) | Student (Pruned) | Student (500 samples) | Final Performance |
|:-------|:-------------------|:-----------------|:---------------------|:------------------|
| **Parameters** | 1,235,814,400 | 1,117,171,392 | 1,117,171,392 | **Maintained** |
| **arc_easy** (acc) | 0.6600 | 0.4800 | **0.5800** | **87.9%** of original |
| **winogrande** (acc) | 0.6000 | 0.5400 | 0.5500 | **91.7%** of original |
| **boolq** (acc) | 0.6700 | 0.7000 | **0.5400** | **80.6%** of original |
| **lambada_openai** (acc) | 0.5700 | 0.1700 | 0.1800 | **31.6%** of original |
